api/src/main/java/graphql/nadel/instrumentation/parameters/NadelInstrumentationTimingParameters.kt

```diff
@@ -0,0 +1,25 @@
+package graphql.nadel.instrumentation.parameters
+
+import graphql.execution.instrumentation.InstrumentationState
+
+data class NadelInstrumentationTimingParameters(
+    val step: Step,
+    private val context: Any?,
+    private val instrumentationState: InstrumentationState?,
+) {
+    fun <T> getContext(): T? {
+        @Suppress("UNCHECKED_CAST") // trust the caller
+        return context as T
+    }
+
+    fun <T : InstrumentationState?> getInstrumentationState(): T? {
+        @Suppress("UNCHECKED_CAST") // trust the caller
+        return instrumentationState as T?
+    }
+
+    enum class Step {
+        ExecutionPlanning,
+        QueryTransforming,
+        ResultTransforming,
```

Comment:
What more should we time? hydrations? renames? Though trick is these don't all run at once (i.e. there's multiple per query etc.), so we'll likely get a bunch of 1ms timings which may be redundant. I could try to combine collate the times then emit them later.

Progress over perfection though, for now I think this will give us some decent metrics to start with.

Reply:
I think this is a good start - however I think we will want some "sub category"

eg.  entering QueryTransforming + CoerceTransform - maybe its just a  class we can use and hence we can metricize on class name

```
tags
   step : QueryTransforming
   stepClass : CoerceTransform

```

Reply:
Thinking on this more - perhaps the `stepName` should just be a string.  I know its not type safe BUT its extensible.  I was thinking class would be ok BUT imagine  1 class with 2 important inner steps that we want to know about.

Perhaps we use type alias to give us type safety onto strings

Reply:
I don't want to use Strings. I think we should be able to easily know based on this enum what we're recording.

That way the caller can do extra things like add extra tags etc based on what is metric is being emitted.

As for sub steps, we can always just encode it in the step name e.g. `CoerceTransform_Query` or `CoerceTransform_Result` etc.

Reply:
I get your desired not the use Strings - but enums (and a sigle set of them) dont compose well. Why not classes then for the sub category.  We can then declare new steps say and it becomes

` step: QueryTransform class : NadelHydrationTransform` comes in on the call back

That way we can know what classes are the fastest / slowest / we can add it up in aggregate if we want